nasa/ow_simulator
# Oceanwater 653 add the ability to apply additional load on the battery at runtime

## Linked Issues:
| EPIC ⚡| [OCEANWATER-84 / Power System Modelling](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-84) |
| :----------- | :----------- |
| Jira Ticket 🎟️  | [OCEANWATER-653 / Add the ability to apply additional load on the battery at runtime ](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-653) |
| Github :octocat:  | N/A |


## Summary of Changes
* Load a csv file representing a given faulty pattern
* Apply the fault at run time while the fault is activated
  - This is currently implemented as an addition to currently generated estimates

## Test
1) Launch the simulation
```bash
roslaunch ow atacama_y1a.launch
```

2) In another terminal with oceanwaters workspace sourced, monitor the different power system output
```bash
rostopic echo /power_system_node/state_of_charge
rostopic echo /power_system_node/remaining_useful_life
rostopic echo /power_system_node/battery_temperature
```

3) Enable any of three different power faults from RQT Dynamic Reconfigure
* low_state_of_charge_power_failure
* instantaneous_capacity_loss_power_failure
* thermal_power_failure

> **Note**: Currently there is no difference in the behavior of the three different faults since the three faults are using the same profile. However, enabling multiple faults at the same time would accumulate the total fault value applied to power, voltage and temperature. 

4) Observe that the values of state_of_charge starts to drop significantly

> **Note:** If the total power applied to power exceed a certain limit (this happens when combining multiple power faults at the same time), then some bad things happen and the power system node crashes at some point.
